llvm/test/CodeGen/MIR/X86/inline-asm-rm-exhaustion.mir

```diff
@@ -0,0 +1,232 @@
+    MOV32mr %stack.0, 1, $noreg, 0, $noreg, %0 :: (store (s32) into %ir.1)
+    $eax = COPY %0
+    RET 0, $eax
+
+...
+---
+name:            inout
+alignment:       16
+exposesReturnsTwice: false
+legalized:       false
+regBankSelected: false
+selected:        false
+failedISel:      false
+tracksRegLiveness: true
+hasWinCFI:       false
+callsEHReturn:   false
+callsUnwindInit: false
+hasEHCatchret:   false
+hasEHScopes:     false
+hasEHFunclets:   false
+isOutlined:      false
+debugInstrRef:   false
+failsVerification: false
+tracksDebugUserValues: false
+registers:
+  - { id: 0, class: gr32, preferred-register: '' }
+  - { id: 1, class: gr32, preferred-register: '' }
+liveins:         []
+frameInfo:
+  isFrameAddressTaken: false
+  isReturnAddressTaken: false
+  hasStackMap:     false
+  hasPatchPoint:   false
+  stackSize:       0
+  offsetAdjustment: 0
+  maxAlignment:    4
+  adjustsStack:    false
+  hasCalls:        false
+  stackProtector:  ''
+  functionContext: ''
+  maxCallFrameSize: 4294967295
+  cvBytesOfCalleeSavedRegisters: 0
+  hasOpaqueSPAdjustment: false
+  hasVAStart:      false
+  hasMustTailInVarArgFunc: false
+  hasTailCall:     false
+  localFrameSize:  0
+  savePoint:       ''
+  restorePoint:    ''
+fixedStack:
+  - { id: 0, type: default, offset: 0, size: 4, alignment: 16, stack-id: default,
+      isImmutable: false, isAliased: false, callee-saved-register: '',
+      callee-saved-restored: true, debug-info-variable: '', debug-info-expression: '',
+      debug-info-location: '' }
+stack:           []
+entry_values:    []
+callSites:       []
+debugValueSubstitutions: []
+constants:       []
+machineFunctionInfo: {}
+body:             |
+  bb.0 (%ir-block.1):
+    ; CHECK-LABEL: name: inout
+    ; CHECK: [[MOV32rm:%[0-9]+]]:gr32 = MOV32rm %fixed-stack.0, 1, $noreg, 0, $noreg :: (load (s32) from %fixed-stack.0, align 16)
+    ; CHECK-NEXT: MOV32mr %stack.0, 1, $noreg, 0, $noreg, [[MOV32rm]] :: (store (s32) into %stack.0)
+    ; CHECK-NEXT: INLINEASM &"# $0 $1", 24 /* mayload maystore attdialect */, 262190 /* mem:m */, %stack.0, 1, $noreg, 0, $noreg, 262190 /* mem:m */, %stack.0, 1, $noreg, 0, $noreg, 12 /* clobber */, implicit-def dead early-clobber $ax, 12 /* clobber */, implicit-def dead early-clobber $cx, 12 /* clobber */, implicit-def dead early-clobber $dx, 12 /* clobber */, implicit-def dead early-clobber $si, 12 /* clobber */, implicit-def dead early-clobber $di, 12 /* clobber */, implicit-def dead early-clobber $bx, 12 /* clobber */, implicit-def dead early-clobber $bp :: (store (s32) into %stack.0)
```

Comment:
I wonder how the extra info like `(load (s32) from %fixed-stack.0, align 16)` etc. being kept in the inlineasm when we have more than one operands? I saw we only kept `(store (s32) into %stack.0)` here.
Besides, we don't have a mem to mem instrcution on X86. Will it be lowered into two instructions when both operands are mem?

Reply:
> I wonder how the extra info like (load (s32) from %fixed-stack.0, align 16) etc. being kept in the inlineasm when we have more than one operands? I saw we only kept (store (s32) into %stack.0) here.

Yeah, it is weird that for an input+output parameter it is printed as just a store (even though the extra info says both `mayload`, `maystore`.  Let me do some debugging of the print method and see if I can give you a more precise answer FWIW.

---

> Besides, we don't have a mem to mem instrcution on X86. Will it be lowered into two instructions when both operands are mem?

heh, we just had [that exact discussion on the parent patch](https://github.com/llvm/llvm-project/pull/70743#discussion_r1392342980) to this one.

In that case, the programmer was wrong to use "rm" for more than one operand, and should have used "r" (regardless if they are using clang or not, with or without my patches).  Also, inline asm may contain multiple instructions, where the inputs/outputs to the inline asm are used as operands for different instructions, not the same instruction which would expose them to the issue you allude to if there was register pressure.

Reply:
> Let me do some debugging of the print method and see if I can give you a more precise answer FWIW.

I believe it's a loop in MachineInstr::print that's iterating the memoperands() of the MachineInstr and calling print on them.

I'm not super familiar with the MachineMemOperand class. Let me triple check whether that loop is indeed the one printing the load/store, and if so, do I need to modify the memoperands() of the MachineInstr (beyond setting the extra info which is very specific to INLINEASM).

Reply:
> I believe it's a loop in MachineInstr::print that's iterating the memoperands() of the MachineInstr and calling print on them.

Ah, close, but it's the other mir printer: `MIPrinter::print`.  So I wonder if I'm missing setting a `memoperands()` when doing the fold...let me look into that.

Reply:
Perhaps I need to add calls to `MachineInstr::addMemOperand`...

Reply:
oh! the caller `TargetInstrInfo::foldMemoryOperand` IS calling `MachineInstr::addMemOperand`, but IIUC it's only doing so once, which is a problem for in/out parameters since technically there's 2 distinct operands, let me work on fixing that up.

Reply:
`MachineInstr::mayFoldInlineAsmRegOp` has a bug with tied operands.

Reply:
fixed in https://github.com/llvm/llvm-project/pull/70832/commits/e553a3761ee40cef80b825b33c820011ca472325 , PTAL